packages/eslint-plugin/tests/rules/no-unnecessary-type-parameters.test.ts

```diff
@@ -167,6 +167,9 @@ ruleTester.run('no-unnecessary-type-parameters', rule, {
         };
       }
     `,
+    'function fn<T extends unknown[]>(...args: T) {}',
```

Comment:
Oh 😕 maybe I was wrong to suggest _any_ type parameter in https://github.com/typescript-eslint/typescript-eslint/issues/9890#issuecomment-2321307729... I think this might be pretty directly a case where the rule should report.

Perhaps a better delineation would be rest type parameters that are used in a return statement?

* Not ok: `declare function fn<T extends unknown[]>(...args: T): void`
* Ok: `declare function fn<T extends unknown[]>(): (...args: T) => void`

Reply:
Yeah, this rule is very controversial! :slightly_smiling_face: 

> rest type parameters that are used in a return statement?

Hmmm, could you please explain why

```ts
declare function fn<T extends unknown[]>(): (...args: T) => void
```

should be considered valid? Maybe I'm missing something


The type `T` cannot be inferred from the type returned by the function, so this function is intended to be used with explicitly specified type arguments?

Looks like a continuation of the https://github.com/typescript-eslint/typescript-eslint/pull/9536#discussion_r1713663117 thread